awx_collection/plugins/lookup/tower_schedule_rrule.py

```diff
@@ -0,0 +1,235 @@
+        description:
+          - The frequency of the schedule
+          - none - Run this schedule once
+          - minute - Run this schedule ever x minutes
+          - hour - Run this schedule every x hours
+          - day - Run this schedule ever x days
+          - week - Run this schedule weekly
+          - month - Run this schedule monthly
+        required: True
+        choices: ['none', 'minute', 'hour', 'day', 'week', 'month']
+      start_date:
+        description:
+          - The date to start the rule
+          - Used for all frequencies
+          - Format should be YYYY-MM-DD [HH:MM:SS]
+        type: str
+      timezone:
+        description:
+          - The timezone to use for this rule
+          - Used for all frequencies
+          - Format should be as US/Eastern
+          - Defaults to America/New_York
+        type: str
+      every:
+        description:
+          - The repition in months, weeks, days hours or minutes
+          - Used for all types except none
+        type: int
+      end_on:
+        description:
+          - How to end this schedule
+          - If this is not defined this schedule will never end
+          - If this is a positive integer this schedule will end after this number of occurances
+          - If this is a date in the format YYYY-MM-DD [HH:MM:SS] this schedule end after this date
+          - Used for all types except none
+        type: str
+      on_days:
+        description:
+          - The days to run this schedule on
+          - A comma seperated list which can contain values sunday, monday, tuesday, wednesday, thursday, friday
+          - Used for week type schedules
+      month_day_number:
+        description:
+          - The day of the month this schedule will run on (0-31)
+          - Used for month type schedules
+          - Can not be used with on_the parameter
+        type: int
+      on_the:
+        description:
+          - A description on when this schedule will run
+          - Two strings seperated by space
+          - First string is one of first, second, third, fourth, last
+          - Second string is one of sunday, monday, tuesday, wednesday, thursday, friday
+          - Used for month type schedules
+          - Can not be used with month_day_number parameters
+    extends_documentation_fragment: awx.awx.auth
+"""
+
+EXAMPLES = """
+    - name: Create a string for a schedule
+      debug:
+        msg: "{{ query('awx.awx.tower_schedule_rrule', 'none', start_date='1979-09-13 03:45:07') }}"
+"""
+
+RETURN = """
+_raw:
+  description:
+    - String in the rrule format
+  type: string
+"""
+
+from ansible.plugins.lookup import LookupBase
+from ansible.errors import AnsibleError
+from datetime import datetime
+import re
+
+missing_modules = []
+try:
+    import pytz
+except ImportError:
+    missing_modules.append('pytz')
+
+try:
+    from dateutil.rrule import rrule, MONTHLY, WEEKLY, DAILY, HOURLY, MINUTELY, MO, TU, WE, TH, FR, SA, SU
```

Comment:
I worry about the complexity of content in the `try` for import exceptions like this. If just _one_ of these items moves within the library, then it will begin failing with the wrong message. Tracebacks aren't great for users to go, but they are light years better than _wrong_ information.

Because of that, I would suggest doing a simplified import of `from dateutil import rrule as rrule`, and then referencing these things like `rrule.MONTHLY` in code. There are other options too. Instead of making these imports here, you can do a _test_ import of `import dateutil` and then import those things after the try-except (again, avoids flagrantly wrong error messages).

Reply:
Fixed.